### PR TITLE
adding a utility to generate an x509 certificate from a private key

### PR DESCRIPTION
**Pull request by vishalchangrani**
Part 1 of adding GRPC secure server support to the access, collection and execution nodes.

### CODE REVIEW

File: network/p2p/keyTranslator.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ func setPubKey(c elliptic.Curve, x *big.Int, y *big.Int) *goecdsa.PublicKey {
 // These utility functions convert a Flow crypto key to a LibP2P key (Flow --> LibP2P)
 
 // privKey converts a Flow private key to a LibP2P Private key
-func privKey(fpk fcrypto.PrivateKey) (lcrypto.PrivKey, error) {
+func PrivKey(fpk fcrypto.PrivateKey) (lcrypto.PrivKey, error) {
```
Comment (vishalchangrani) [inline network/p2p/keyTranslator.go:38]: making it public since this is now used for certificate generation as well.

File: utils/grpc/grpc.go [hunk 2]
```diff
@@ -1,5 +1,55 @@
...
+	// extract the TLSConfig from it which will contains the generated x509 certificate
+	libp2pTlsConfig, _ := id.ConfigForAny()
+
+	// verify that exactly one certificate was generated for the given key
+	if len(libp2pTlsConfig.Certificates) == 0 {
+		return nil, fmt.Errorf("failed to generate the x509 certificate")
+	}
+	if len(libp2pTlsConfig.Certificates) > 1 {
+		return nil, fmt.Errorf("unexpected number of x509 certificate chains generated")
+	}
```
Comment (huitseeker) [inline utils/grpc/grpc.go:43]: a `len(libp2pTlsConfig.Certificates) != 1` might suffice, but ymmv.

File: network/p2p/keyTranslator.go (hunk 1 above)
Comment (huitseeker) [reply network/p2p/keyTranslator.go:38]: Not quite your PR, but should we return an error if the signing algorithm is not matched within the set of unmarshallers? e.g. `crypto.BLSBLS12381` ?

**Review by huitseeker — APPROVED** [review]
Thanks!

File: network/p2p/keyTranslator.go [hunk 3]
```diff
@@ -35,7 +35,7 @@ func setPubKey(c elliptic.Curve, x *big.Int, y *big.Int) *goecdsa.PublicKey {
 // These utility functions convert a Flow crypto key to a LibP2P key (Flow --> LibP2P)
 
 // privKey converts a Flow private key to a LibP2P Private key
```
Comment (smnzhu) [inline network/p2p/keyTranslator.go:37]:
```suggestion
// PrivKey converts a Flow private key to a LibP2P Private key
```

File: utils/grpc/grpc.go [hunk 4]
```diff
@@ -1,5 +1,55 @@
...
+
+	// create a libp2p Identity from the libp2p private key
+	id, err := libp2ptls.NewIdentity(libP2PKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not generate identity: %w", err)
+	}
+
+	// extract the TLSConfig from it which will contains the generated x509 certificate
+	libp2pTlsConfig, _ := id.ConfigForAny()
```
Comment (smnzhu) [inline utils/grpc/grpc.go:35]: What is the second return value here?

**Review by smnzhu — APPROVED** [review]

File: utils/grpc/grpc.go (hunk 2 above)
Comment (vishalchangrani) [reply utils/grpc/grpc.go:43]: agreed will change

File: utils/grpc/grpc.go (hunk 4 above)
Comment (vishalchangrani) [reply utils/grpc/grpc.go:35]: It is the public key of the private key that was used to generate the ID. This `ConfigForAny` takes an ID and returns the TLS config and the public key for the ID.

File: utils/grpc/grpc.go [hunk 5]
```diff
@@ -1,5 +1,54 @@
...
+		return nil, fmt.Errorf("could not convert Flow key to libp2p key: %w", err)
+	}
+
+	// create a libp2p Identity from the libp2p private key
+	id, err := libp2ptls.NewIdentity(libP2PKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not generate identity: %w", err)
+	}
+
+	// extract the TLSConfig from it which will contains the generated x509 certificate
```
Comment (tarakby) [inline utils/grpc/grpc.go:34]:
```suggestion
	// extract the TLSConfig from it which will contain the generated x509 certificate
```

**Review by tarakby — APPROVED** [review]
I had a look at the libp2p and tls functions used and it looks good to me 👌

**Comment by vishalchangrani** [comment]
> I had a look at the libp2p and tls functions used and it looks good to me 👌

❤️ thanks

**Pull request merged**